### ISSUE TITLE
Question: pathlib compatibility?

### ISSUE THREAD

**Issue by smheidrich** [PyFilesystem/pyfilesystem2#238]
Are there any plans to make the functionality offered by this package available through an interface mirroring the Python standard library's [``pathlib.Path``](https://docs.python.org/library/pathlib.html), or does anyone know of external projects working towards this? I think adoption of ``pathlib`` is already comparatively widespread, so if something like this existed, a decent number of projects would obtain support for the various filesystems offered by PyFilesystem with no or minimal changes.

**Comment by smheidrich** [comment]
The way I see it, implementing the ``PathLike`` interface (``__fspath__``) would be pretty pointless, because its only purpose is to return the path as a string so that it can be used in the parts of the standard library which are mostly just wrappers around syscalls (``open``, ``os.mkdir``, ...). This wouldn't be useful when dealing with any of the interesting filesystems supported by PyFilesystem (remote, in-memory, ...).

But what I meant was the ``pathlib.Path`` interface, which defines methods for (almost) anything one needs when dealing with files, such as [``Path.open``](https://docs.python.org/3/library/pathlib.html#pathlib.Path.open), [``Path.iterdir``](https://docs.python.org/3/library/pathlib.html#pathlib.Path.iterdir), and all the methods in the right column [here](https://docs.python.org/3/library/pathlib.html#correspondence-to-tools-in-the-os-module). Not all of these methods make sense for all of the filesystems supported by PyFilesystem, but I'd imagine the most commonly used ones are things like ``open``, ``iterdir``, ``exists``, ``mkdir``, ``rename``, ... which map nicely onto [``fs.base.open``](https://docs.pyfilesystem.org/en/latest/reference/base.html#fs.base.FS.open), [``fs.base.listdir``](https://docs.pyfilesystem.org/en/latest/reference/base.html#fs.base.FS.listdir), [``fs.base.exists``](https://docs.pyfilesystem.org/en/latest/reference/base.html#fs.base.FS.exists), [``fs.base.makedir``](https://docs.pyfilesystem.org/en/latest/reference/base.html#fs.base.FS.makedir), [``fs.base.move``](https://docs.pyfilesystem.org/en/latest/reference/base.html#fs.base.FS.move), ... you get the idea.

So if there were classes implementing the various ``pathlib.Path`` methods, but internally passing them on to corresponding PyFilesystem methods, any library which only performs file operations using ``pathlib.Path`` objects (e.g. ``path.open()`` instead of ``open(path)``) would automatically support all of PyFilesystem's supported filesystems (while using only a limited subset of its full functionality, of course). All the user of the library would have to do is to create and pass in instances of the aforementioned classes.

There are some issues with this idea, e.g. [``pathlib.Path`` doesn't have a ``copy`` method](https://bugs.python.org/issue24229) and so I imagine most people using ``pathlib`` fall back to something like [``shutil.copy``](https://docs.python.org/3/library/shutil.html#shutil.copy) (which internally does ``open(path)`` ⇒ abstraction broken), but I have some ideas for how to sort this out.

Anyway, I could throw together a library for this myself, I just didn't want to reinvent the wheel in case someone had already done it.

**Comment by willmcgugan** [comment]
I'm not a big fan of the abstraction offered by pathlib, but of course I'm a little biased. Personally if I had an FS object I'd opt to use that over a path object. But tastes vary, perhaps pathlib just fits others brains better.

@althonos suggested making all pyfilesystem objects support a pathlike interface, which I think is more or less what you are suggesting @smheidrich 

If you want to tackle that yourself, I'd certainly be happy to advise.

**Comment by smheidrich** [comment]
@althonos @willmcgugan Thanks for the help, I might write something up myself then. :+1:

**Comment by smheidrich** [comment]
@CMCDragonkai It's probably best to open a new issue about this, I think it's different from what I was talking about here.

**Comment by willmcgugan** [comment]
@CMCDragonkai Mixing PyFIlesystem with pathlib is generally not a good idea (with a few possible exceptions). A path generated by a Path object may work with OSFS but might not work with any of the other filesystems.

In general, you should use `fs.path` to manipulate paths. The library will ensure that the underlying storage receives the correctly formatted path. See https://docs.pyfilesystem.org/en/latest/concepts.html#paths for details.

**Comment by arogozhnikov** [comment]
> I'm not a big fan of the abstraction offered by pathlib, but of course I'm a little biased. Personally if I had an FS object I'd opt to use that over a path object. But tastes vary, perhaps pathlib just fits others brains better.

let me try to convince you that pathlib-like interface for fs2 is a good idea. 

Main pro is you can point to object/folder by passing a single object. 
If I need to pass a folder to a function, Path object alone is sufficient. In case of fs2, I need to pass both fs and path inside;

Same with passing files. There are cases when data consists of main file and a set of optional additional files that only have a different suffix. Exampe from my current area is genomics files (e.g. foo.bam can be accompanied by an index file foo.bam.bai), so passing reference to file should always be 'with path' not 'with stream'. Passing Path allows easily getting both and checking if index file is present.

I am new to fs2, feel free to correct me if I'm missing something

**Comment by lurch** [comment]
I _think_ you might be misunderstanding the abstraction that pyfilesystem2 provides @arogozhnikov ?

AIUI, "Path object" is effectively just a wrapper around an underlying filepath string? (i.e. a filepath that exists in your on-disk filesystem.) But as @willmcgugan mentioned earlier, pyfilesystem also provides abstractions around filepaths that _don't_ exist in your on-disk filesystem, e.g. [FTPFS](https://docs.pyfilesystem.org/en/latest/reference/ftpfs.html) or [MemoryFS](https://docs.pyfilesystem.org/en/latest/reference/memoryfs.html).
I don't think there's any way a "single Path object" could represent a file on a remote FTP server for example?

**Comment by arogozhnikov** [comment]
> I don't think there's any way a "single Path object" could represent a file on a remote FTP server for example?

trivial - it just keeps 1) reference to fs object and 2) path within that fs internally

> I think you might be misunderstanding the abstraction that pyfilesystem2 provides

I don't think so

**Comment by lurch** [comment]
> trivial - it just keeps 1) reference to fs object and 2) path within that fs internally

Well in that case it'd be a custom object, containing both a pyfilesystem FS **and** a path. Which is something very different to the `Path` object described [here](https://docs.python.org/3/library/pathlib.html).

Maybe if you gave an example of what it is you're trying to do, that might help clarify things?

**Comment by arogozhnikov** [comment]
Topic-starter has a nice wording:
> an interface mirroring the Python standard library's pathlib.Path

that is - custom objects (in your words) that expose interface identical to `pathlib.Path` but can operate on fs2 filesystems

**Comment by smheidrich** [comment]
FYI, I started writing an interface [here](https://github.com/smheidrich/pathlib-fs), although it's still missing a lot of the less frequently used `Path` methods because I ended up not needing this whole thing after all.

It also became clear that the idea might be a bit ill-defined regarding methods like `Path.absolute()` (to get the "true" absolute path of something on an OSFS, you'd have to read the OSFS's root path, which goes against FS-agnosticism) or `Path.drive()` (same issue: only really makes sense to have for OSFS).

Another issue could be serialization: `Path`s can be serialized and existing code using pathlib might just rely on that - attaching an FS object to the path breaks that property.

But I guess neither of these are show-stopping issues.
